### ISSUE TITLE
wamp.ApplicationSession examples should call super()

### ISSUE THREAD

**Issue by terrycojones** [tavendo/AutobahnPython#457]
The docs show some examples of using `wamp.ApplicationSession`, e.g., at https://github.com/tavendo/AutobahnPython/blob/master/examples/asyncio/wamp/rpc/timeservice/frontend.py

Neither of the two methods (`onJoin` and `onDisconnect`) calls the equivalent method in the super class. That happens to be ok (at least for now) because the super class implementations don't do anything. But the examples can lead the reader to forget about / ignore the super class. So, for example, a beginner might write their own `onLeave` and not call `super().onLeave()` with just a `print` statement in it. That results in the transport never being disconnected.

I think things would be clearer if the super methods were called, even though they don't currently do anything.

Unfortunately if you call `super.onChallenge` though, you get an exception, which is not logged to the terminal :-(  So it's not always a good idea to call the super method. But hopefully by the time people are writing their own authenticators they'll have more of an idea what they're doing.

**Comment by meejah** [comment]
There should definitely be clear documentation on when/if you need to call the super-class methods. ``onChallenge`` does need to be overridden currently if you do authentication -- but again, that should be made clear in the docs.

For ``onLeave``: it should be documented what the default implementation does, and the user can choose appropriately. At present, overriding ``onLeave`` to *not* call ``.disconnect`` is the only way you can achieve the "leave, and re-join with the same transport" use-case that I know of.

So, I *believe* it's correct to say the the ``on*()`` methods are intended to be overridden without calling ``super()``. Right @oberstet? If so, this bug should be closed when the docs are updated properly to reflect that.

**Comment by oberstet** [comment]
> So, I believe it's correct to say the the on*() methods are intended to be overridden

Yes. And the same does not hold true for the ctor. It's docs missing ..

**Comment by meejah** [comment]
Also so we don't miss it: it should be pointed out in the docs that if you *do* override ``onLeave`` you probably want to call ``self.disconnect()`` if you want the transport to go away when the WAMP session is over.

**Comment by oberstet** [comment]
In summary, the docs should cover:

- always call the base ctor when overriding the ctor
- all callbacks may be overridden without calling the base implementation
- the base implementation for on_xxx does XXX
- if you override on_xxx, consider doing XXX (like calling disconnect() in on_leave())
